sklearn/preprocessing/_function_transformer.py

```diff
@@ -72,18 +104,55 @@ class FunctionTransformer(BaseEstimator, TransformerMixin):
         Dictionary of additional keyword arguments to pass to inverse_func.
 
     """
-    def __init__(self, func=None, inverse_func=None, validate=True,
-                 accept_sparse=False, pass_y='deprecated', check_inverse=True,
-                 kw_args=None, inv_kw_args=None):
+    def __init__(self, func=None, inverse_func=None, validate=None,
+                 accept_sparse=False, force_all_finite=True,
+                 pass_y='deprecated', check_inverse=True, kw_args=None,
+                 inv_kw_args=None):
         self.func = func
         self.inverse_func = inverse_func
         self.validate = validate
         self.accept_sparse = accept_sparse
+        self.force_all_finite = force_all_finite
         self.pass_y = pass_y
         self.check_inverse = check_inverse
         self.kw_args = kw_args
         self.inv_kw_args = inv_kw_args
 
+    def _check_input(self, X):
+        # FIXME: Future warning to be removed in 0.22
+        if self.validate is None:
+            self._validate = True
+            if hasattr(X, 'loc'):
+                warnings.warn("The default validate=True will be replaced by "
+                              "validate='array-or-frame' in 0.22. A pandas "
+                              "DataFrame will not be converted to a 2D "
+                              "NumPy array.", FutureWarning)
+        else:
+            self._validate = self.validate
+
+        if ((not isinstance(self._validate, bool)) and
+                self._validate != 'array-or-frame'):
+            raise ValueError("'validate' should be a boolean or "
+                             "'array-or-frame'. Got {!r} instead."
+                             .format(self._validate))
+        if ((not isinstance(self.force_all_finite, bool)) and
+                self.force_all_finite != 'allow-nan'):
+            raise ValueError("'force_all_finite' should be a boolean "
+                             "or 'allow-nan'. Got {!r} instead."
+                             .format(self.force_all_finite))
+
+        if self._validate:
+            if hasattr(X, 'loc') and self._validate == 'array-or-frame':
```

Comment:
Actually that might allow people to pass xarray data array and it should even work. The validation for the nan and inf should is currently working for those arrays. Then, we have the following solutions:

- Let the ducktyping and do not document that xarray is supported;
- Let the ducktyping and say that we can pass xarray.
- Make a strict isinstance.

@rth WDYT?

Reply:
Interesting.

> Let the ducktyping and do not document that xarray is supported;

Let's keep things as they are now, then. The second solution would require some more unit tests (and optionally a CI build with xarray) and since pandas support is only in progress, I don't think it would be reasonable to officially support yet another data format at this time..